### PR TITLE
Thyra epetra adapters refactor

### LINKED ISSUE #3: Epetra_Map -> create_VectorSpace() -> get_Epetra_Map() doesn't return the original map
Given an Epetra_Map, one can convert it to a Thyra::VectorSpaceBase,

``` c++
Teuchos::RCP<const Thyra::VectorSpaceBase<double>> space =
    Thyra::create_VectorSpace(mymap);
```

When translating the result of this back to an Epetra_Map,

``` c++
Teuchos::RCP<const Epetra_Map> spacemap =
    Thyra::get_Epetra_Map(*space, comm);
```

it's not the same as the original map.

``` c++
spacemap->SameAs(*mymap) // false
```

(From [Sandia bug 6341](https://software.sandia.gov/bugzilla/show_bug.cgi?id=6341).)

### PR DESCRIPTION
**Pull request by bartgol**
@trilinos/thyra
@trilinos/epetra

## Description

This PR introduces new interfaces for the Thyra-Epetra adapters, so that they are the analogue of the Thyra-Tpetra ones. In particular, the new interfaces act as 'wrappers', and the conversion Epetra->Thyra->Epetra is indeed an identity map. This addresses issue #3 . The new interfaces are not enabled by default. To enable them, one needs to set ThyraEpetraAdapters_ENABLE_Refactor to ON in the cmake configure script. If not enabled, the old interfaces are still available.

Note: I fixed as much code as I could, to work with the new interfaces. However, I didn't enable the whole Trilinos. I will fix issues as the autotester brings them up. It is worth notice that places that use Thyra-Epetra interfaces are likely filled with `#ifdef HAVE_THYRA_EPETRA_REFACTOR` guards, to switch between old and new interface. This is not really good looking. I tried to keep some interfaces the same as the old ones, even though I do not fully understand (or agree with) them. For instance in the LinearOp, whose initialization method takes supportsAdjoint as input (while I would let the LinearOp verify whether or not it is supported, by testing transpose support in the wrapped Epetra op). Also other interfaces have the same signature, even though the naming convention differs from the rest of the subpackage. E.g., `Thyra::get_Epetra_Opertor` in `Thyra_get_Epetra_Operator.hpp` can be easily replaced by `Thyra::EpetraOperatorVectorExtraction::get(Const)EpetraOperator`, but keeping it has allowed to lower the amount of places where I needed an ifdef for the new implementation. Ideally, I would remove them later on though.

## Motivation and Context

The root of this PR is in issue #3 . See discussion therein for more details.

## Related Issues

* Closes #3 

## How Has This Been Tested?

I added a testsuite for the new interfaces, by adapting the testsuite found  for the tpetra adapters.

## Checklist

- [x] My commit messages mention the appropriate GitHub issue numbers.
- [x] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

## Additional Information
The new interfaces require a new cmake option to be enabled. I am not sure how the autotester can handle this.

There is one test currently failing in Stratimikos. I'm chasing it, but I wanted to put the PR up so I could start getting feedback.

### CODE REVIEW

**Comment by bartgol** [comment]
Sorry. I selected the wrong base branch. Closing.

**Comment by bartgol** [comment]
I didn't know you could change the base branch. I'm reopening the PR now that it is based on develop.

**Comment by bartgol** [comment]
The failing test in Stratimikos is `Stratimikos_test_single_belos_thyra_solver_driver_nos1_nrhs8_MPI_1`, and fails due to the following:

  
```
  Check: rel_err(sum(v4),sum(v5)) <= linear_properties_error_tol() ?
     ...
     3: rel_err(-9.08147e+06,-9.08147e+06) = 1.3783e-13 <= 1e-13 : FAILED
     ...
     7: rel_err(5.53999e+06,5.53999e+06) = 2.80406e-13 <= 1e-13 : FAILED

```

I will dig further.

**Comment by mhoemmen** [comment]
The only issue with PR testing is that it won't exercise the case where your code is actually enabled.  You could first enable your code by default, make sure the PR tests pass, then add another commit that disables it, and check the PR tests again.

**Comment by bartgol** [comment]
@mhoemmen Ok, I will push a commit enabling it by default.

**Comment by bartlettroscoe** [comment]
I also have @mhoemmen's concern about this being disabled by default (and at configure-time at that).

If we have any hope of deprecating the old implementation, I think that both implementations need to be available and tested in the default configuration of Trilinos.  It should be a runtime decision as to which implementation one wants to use.   I think some helper functions, some conditional dynamic casts, etc. can make this work very smoothly.  And I think that we can make the current direct interface functions emit deprecated compile-time warnings so that users of Trilinos.  That will allow for testing for both approaches and a smoother transition to the new implementation both inside of Trilinos and in customer codes.

If you can give me a SNL P/T (offline), then I can spend some time looking into updating this branch to accomplish those goals.

**Comment by mhoemmen** [comment]
@bartgol wrote:
> But perhaps it would be nice to get a first run of the autotester that checks the existing version (with the feature not enabled) is still clean?

Yup (ok to do the tests in that order :) )

**Comment by bartgol** [comment]
@bartlettroscoe I am not sure if compiling both at the same time and leave just a runtime choice is feasible/ideal. Two reasons:

- there is one class (EpetraLinearOp) with the same name in the two implementations, but different class interface. I did it because a) it made its interface 'equivalent' to that of the Tpetra adapters, b) some things (like partiallyInitialize) made no sense to me (and were never used inside Trilinos). For this class, I need to know which interface I am using, in order to include the correct file.
- old interfaces in Thyra_EpetraThyraWrappers require to pass an Epetra_Map (or comm) alongside the Thyra object, in order to retrieve the Epetra object. This is imho defying the wrapper purpose. I want to get the epetravector to feed it to something that needs an epetravector, but I don't want to (or can't) keep around an epetra map. Nor I want to have to retrieve it everytime I need to get the epetra stuff from the thyra stuff.

One potential solution would be (though it may break downstream code) to rename the old interface file something like Thyra_EpetraLinearOpDeprecated.hpp", and add deprecated interfaces to the new EpetraLinearOp. Also, I could add a `Thyra_EpetraThyraWrappersDeprecated.hpp` file which wraps calls to the new interface inside old interfaces (issuing deprecation). That way you would not have to check any macro to switch between implementations. But if you wrap the new interface inside a the old interfaces, you may as well remove all the old interfaces from Trilinos... And about that: the old interfaces allowed to create, say, an Epetra vector from a Thyra vector, regardless of whether the Thyra vector was created from an Epetra one or not. I added interfaces (e.g., `getOrCreateEpetraVector`) that try to cast and fall back on copying, but I didn't use them in the downstream packages. It would be nice to understand if downstream packages (e.g., Teko) were trying to support that scenario, or whether they were always assuming Epetra->Thyra->Epetra conversions...

**Comment by bartlettroscoe** [comment]
@bartgol, I would have to look over this more carefully, but I bet an approach like the following would work:

* Put the new interface functions and classes under that `Thyra::New::` namespace so that they don't conflict with the old/existing interface.

* Rename the old interface functions `<current-name>_impl` and and then provide compile-time deprecated wrappers for the old interface function names `<current-name>` to call `<current-name>_impl`.

* Extend the new interface functions and classes to accept Thyra objects created by the old interface functions (just some simply dyn-casting if statements. can do that).  And write automated tests for all of the supported/needed use cases.

* Refactor all of the Trilinos packages to use the new functions/classes under `Thyra::New::`.

I would have to work out the details, but what the above would do is that it would not break backward compatibility at all but yet would provide deprecated warnings to downstream customers warning them to switch to the new interfaces.  Also, it would still provide for some of the existing "magic" conversions that the existing interface currently allows (for the time being).

Then, in a few months (or years), we remove the old duplicated interfaces/adapters and move the new interfaces/adapters from `Thyra::New::` into `Thyra::` and provide a script that clients can use to refactor their code (there are several examples of these scripts you can find under packages/thyra/refactoring/).

We need an incremental approach in order for this to be feasible or it will cause someone a lot of pain at a bad time.

If we were to merge this PR as is, it would get no PR testing and no other automated testing at all in Trilinos.  But since Albany would be using this, Albany would be the first entity to see breakages in this code.  That would not be fun for Albany developer or Trilinos developers (who would get yelled at to fix ASAP which would not be fair).

File: packages/muelu/adapters/stratimikos/Thyra_MueLuPreconditionerFactory_decl.hpp [hunk 1]
```diff
@@ -394,7 +394,11 @@ namespace Thyra {
 #if defined(HAVE_MUELU_EPETRA)// && defined(HAVE_MUELU_SERIAL)
         if (bIsEpetra) {
           RCP<ThyEpLinOp> epetr_precOp = rcp_dynamic_cast<ThyEpLinOp>(thyra_precOp);
+#ifdef HAVE_THYRA_EPETRA_REFACTOR
```
[the remark this answers is not in the document]
Comment (bartgol) [reply packages/muelu/adapters/stratimikos/Thyra_MueLuPreconditionerFactory_decl.hpp]: I agree

**Comment by bartgol** [comment]
@bartlettroscoe Thanks for your suggestions. It is going to take me some time to work on this. I am doing it almost only on my free time. If you guys are ok with having an open PR for days (weeks), great, otherwise, I can close it and reopen it when the work is close to be done, including your suggestions.

**Comment by bartlettroscoe** [comment]
@bartgol, it is fine to leave this open for a long as we want.  There is zero harm in doing that.  I have added the "AT: WIP" label and that will make it clear this is getting worked on and it will avoid auto PR testing iterations until we want.

Thanks for taking the time to work on this!  I will try to help as I can find some time (and some funding).

**Comment by prwolfe** [comment]
This has been inactive for over 16 months.  Should we close this PR or is it still being pursued?

**Comment by bartgol** [comment]
@prwolfe Sorry about that. I am not currently pursuing it, so I guess I can close it. Thanks!

**Comment by prwolfe** [comment]
Cool – Thanks!

**Pull request closed without merging**